Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py

```diff
@@ -0,0 +1,714 @@
+    ON_EDGE = "Peak mask is on the detector edge"
+    NO_PEAK = "No peak detected."
+
+
+class InstrumentArrayConverter:
+    """
+    This class is used to convert the spectra in the region of a peak to a 3D numpy array of signal and errors
+    Note that for RectangularDetectors and GridDetectors it assumes that banks are not adjacent
+    - i.e. it truncates the data to the edge of the bank.  For general ComponentArray type detector banks it does
+     search for banks with adjacent column-components (on WISH this corresponds to tubes).
+    """
+
+    def __init__(self, ws):
+        self.ws = ws
+        self.inst = ws.getInstrument()
+        if any(self.inst[icomp] for icomp in range(self.inst.nelements()) if
+               isinstance(self.inst[icomp], RectangularDetector) or isinstance(self.inst[icomp], GridDetector)):
+            # might not be true for all components due to presence of monitors etc.
+            self.get_detid_array = self._get_detid_array_rect_detector
+        else:
+            self.get_detid_array = self._get_detid_array_comp_assembly
+
+    @staticmethod
+    def _split_string_trailing_int(name):
+        *_, match = re.finditer(r'\d+', name)  # last match
+        return name[:match.start()], name[match.start():]  # prefix, num_str
+
+    @staticmethod
+    def _find_nearest_child_to_component(parent, comp, excl=None, ndepth=1):
+        dist = np.inf
+        nearest_child = None
+        for ichild in range(parent.nelements()):
+            child = parent[ichild]
+            if hasattr(child, 'nelements'):
+                while child.nelements() == 1:
+                    child = child[0]
+            is_correct_type = 'CompAssembly' in child.type() or child.type() == 'DetectorComponent'
+            is_not_excluded = excl is None or child.getFullName() != excl
+            if is_correct_type and is_not_excluded:
+                this_dist = child.getDistance(comp)
+                if this_dist < dist:
+                    dist = this_dist
+                    nearest_child = child
+        if ndepth > 1:
+            nearest_child = InstrumentArrayConverter._find_nearest_child_to_component(nearest_child, comp,
+                                                                                      ndepth=ndepth - 1)
+        return nearest_child
+
+    def _find_nearest_adjacent_col_component(self, dcol, col, ncols, bank_name, col_prefix, col_str, delim, ncols_edge):
+        next_col_prefix, next_col_str = None, None
+        isLHS = np.any(dcol + col <= ncols_edge)
+        isRHS = np.any(dcol + col >= ncols - ncols_edge + 1)
+        if isLHS or isRHS:
+            # get tube/col at each end of detector to get tube separation
+            detLHS = self.inst.getComponentByName(
+                delim.join([bank_name, f'{col_prefix}{1:0{len(col_str)}d}']))
+            detRHS = self.inst.getComponentByName(
+                delim.join([bank_name, f'{col_prefix}{ncols:0{len(col_str)}d}']))
```

Comment:
I think there's an assumption in here that the "module" names end in a number. Perhaps worth documenting this somewhere (unless it's easy to generalise to the case where the module name doesn't end in a number).

I spotted this when running on a daft input workspace for another instrument (POLARIS) where the module names are things like "bank1module", "bank2module" and in this case the detector name passed into `getComponentByName` isn't constructed properly

Reply:
I have assumed the component names (for rows and columns) end with an integer as this is the assumption made in the code which populates the row and column in peak object/workspace (so a def-facto standard for single-crystal diffractometers with component array detector banks? ). 
https://github.com/mantidproject/mantid/blob/bc222196e4303a5ce799455408e412eb4c0b04a2/Framework/DataObjects/src/Peak.cpp#L229-L241
I agree that I should state this in the docs though (also the assumption that the indexing starts from 1 not 0 - i.e. tube001 is the first tube. And probably also throw a warning if no component is found 9as per a previous comment).

Reply:
Perhaps I should strip the component name of any trailing non-digit characters?

Reply:
Just checked on POLARIS (component array banks) and making a `SingleCrystalPeakTable` using pick tool produces peaks with column = -1 (which wouldn't work with this approach anyway). So I will just document the assumptions made about the instrument component names.